src/modm/driver/encoder/encoder_input_impl.hpp

```diff
@@ -16,28 +18,22 @@ modm::BitBangEncoderInput<PH_A, PH_B, DIVISION, UnderlyingType>::update()
 	if (diff & 1)
 	{
 		raw_last = raw;               // store new as next last
-		enc_delta += (diff & 2) - 1;  // bit 1 = direction (+/-)
+		delta += (diff & 2) - 1;  // bit 1 = direction (+/-)
 	}
+
+	using DebugLEDs = modm::platform::SoftwareGpioPort<Board::D5, Board::D6, Board::D7, Board::D8, Board::D9, Board::D10, Board::D11, Board::D12>;
+	DebugLEDs::write(delta);
 }
 
-template<typename PH_A, typename PH_B, modm::bitbang_encoder_input::Division_t DIVISION, std::signed_integral UnderlyingType>
+template<typename PH_A, typename PH_B, uint8_t DIVISION, std::signed_integral UnderlyingType>
 UnderlyingType
-modm::BitBangEncoderInput<PH_A, PH_B, DIVISION, UnderlyingType>::getValue()
+modm::BitBangEncoderInput<PH_A, PH_B, DIVISION, UnderlyingType>::getIncrement()
 {
 	::modm::atomic::Lock _;
-	UnderlyingType val = enc_delta;  // counts since last call
-	switch (DIVISION)
-	{
-		case bitbang_encoder_input::Division::ONE:
-			enc_delta = 0;
-			break;
-		case bitbang_encoder_input::Division::TWO:
-			enc_delta = val & 1;
-			val >>= 1;
-			break;
-		case bitbang_encoder_input::Division::FOUR:
-			enc_delta = val & 3;
-			val >>= 2;
-	}
+	UnderlyingType val = delta;  // counts since last call
+
+	delta = val & (DIVISION - 1);
+	constexpr uint8_t shift = std::log2(DIVISION);
+	val >>= shift;
```

Comment:
Isn't this just `val /= DIVISION`?
Alternatively you can require DIVISION to be a power of two and generate the mask via `(1ul << DIVISION) - 1` and trivially shift.

Reply:
Alternatively you can interpret DIVISION as a power of two...
I think the **interface generalizes better**, when DIVIDER is "number of signal-flips to  count one increment" instead of Power of two thing ;)

Reply:
Isn't this just val /= DIVISION?

My analytic understanding confirms:
- `val >>= log2(DIVISOR)` _equals_ `val /=DIVISION`
- Shifting by log2 results in the division

But `val />> DIVISOR` does not work properly:

- When counting up, it behaves the same like `val >>= log2(DIVISOR)`
- But when counting down, the double amount of pulses is required, to get one decrement

I'm digging for the explanation. A good thing to learn more about twos complement.

Reply:
Found the leak in `var /= DIVISOR` when counting down that was not happening with `val >>= log2(DIVISOR)` when counting down. FYI:

### Let's have

int8_t val = -4;
const uint8_t DIVISOR = 8;

### The difference is the rounding behaviour:

**`var /= DIVISOR` always rounds up.**
```
MODM_LOG_INFO << "val / DIVISOR = " << val / DIVISOR << modm::endl;; // outputs 0
```

**`val >>= log2(DIVISOR)` rounds _"smart"_,  away from 0!**
```
constexpr uint8_t shift = std::log2(DIVISOR);
MODM_LOG_INFO << "val >> log2(DIVISOR) = " << (val >> shift) << modm::endl; // outputs -1
```

Reply:
Ah, of course… signed division is not the same as shifting! Ok, let's leave it like this.

Reply:
Again FYI, Here's why:
signed int has two representations of 0:
0000 -> +0
1111 -> -0

We can say negatives have an offset of -1 against positives.
The compiler and/or ALU knows this and handles the offset when dividing signed ints.
**But shift divisions are dump and therefore round down for negatives.**